File: 수학(구현)/백준_1790_수 이어쓰기2/수 이어쓰기2.cpp
#include <iostream>
#include <cmath>
using namespace std;

int main() {
	int N, k;
	scanf("%d %d", &N, &k);
	
	int norm = 1;
	long long norm_cnt = 9;
	long long all_cnt = 0;
	
	while (N > norm_cnt) {
		N -= norm_cnt;
		all_cnt += (norm * norm_cnt);
		norm++;
		norm_cnt *= 10;
	}
	all_cnt += ((long long)N * norm);

	norm = 1;
	norm_cnt = 9;
	int cnt = 0;

	while (k > cnt + (norm * norm_cnt)) {
		cnt += (norm * norm_cnt);
		norm++;
		norm_cnt *= 10;
	}
	
	int cal = k - (cnt + 1);
	int quotient = cal / norm;
	int remainder = cal % norm;
	int i = norm - remainder;
	int num = pow(10, norm-1) + quotient;
	
	int ans = 0;
	while (i--) {
		ans = num % 10;
		num /= 10;
	}
	
	if (k > all_cnt) printf("-1");
	else printf("%d", ans);

}

File: 수학(구현)/백준_1790_수 이어쓰기2/수 이어쓰기2.kt
import java.util.*

fun main(args: Array<String>) {
    val s = Scanner(System.`in`)
    var N: Long = s.nextLong()
    var k: Long = s.nextLong()

    var norm: Long = 1
    var norm_cnt: Long = 9
    var all_cnt: Long = 0

    while (N > norm_cnt) {
        N -= norm_cnt
        all_cnt += (norm * norm_cnt)
        norm++
        norm_cnt *= 10
    }
    all_cnt += (N * norm)

    norm = 1
    norm_cnt = 9
    var cnt: Long = 0

    while (k > cnt + (norm * norm_cnt)) {
        cnt += (norm * norm_cnt)
        norm++
        norm_cnt *= 10
    }
    var cal: Long = k - (cnt + 1)
    var quotient: Long = cal / norm
    var remainder: Long = cal % norm
    var i: Long = norm - remainder
    var num: Long = Math.pow(10.toDouble(), (norm - 1).toDouble()).toLong() + quotient

    var ans: Long = 0
    while (i-- > 0) {
        ans = num % 10
        num /= 10
    }
    if (k > all_cnt) print("-1") else print(ans)
}
